### ISSUE TITLE
Feature request: split index files, so they don't grow too big

### ISSUE THREAD

**Issue by zcalusic** [restic/restic#1412]
Currently, as the final product of the rebuild-index/prune commands, a single big index is written out. And it can grow really large. On one of my repos, it has grown to 455MB. Such big files are not a problem for local disk repos, FS can handle them just fine. But, when using any kind of remote repo with HTTP transport, it's becoming a problem, as can be seen in #1386, #1411 and possibly more reported issues.

My suggestion would be to split index files, not unlike data blobs, so that there are more smaller size indexes which together form a full index. And my suggestion would also be that single index file is never bigger than about 64MB, but probably we could go for even smaller limit, somewhere in the 16 - 32MB.

And to not open a new issue, I'd also like to discuss the possibility of bringing index to the local side, would that make any sense? If I understand correctly, index files are basically kind of cache files, they are there just to speedup some lookups, right? They don't hold any new information, just information that's already in the repo, but packed in a way that makes it useful to do faster information retrieval about objects in the repo. So, unless I'm overlooking something, they could exist only on the local side, and the repo itself would be still complete?

I understand that increasing usage on the client side would be alarming for some, but even if this is implemented as an option would work. Then 455MB would not need to be uploaded over and over to the remote side, just to be pulled back to be used for the next restic operation. And then we could even consider using some other kind of on-disk fast lookup database (Bolt or similar), which I feel would bring restic memory usage down and make it more usable on smaller systems like Raspberry, various NAS devices etc...?

OK, that was a brain dump, it feels rather logical, but I'm probably missing something...

**Comment by fd0** [comment]
Definitely a good idea. The index file format can be split easily without breaking backwards compatibility, and I think we should try that.

The index files consist of all the headers of the pack files (basically "which blob is stored where"). This prevents having to list the pack files and read all headers of all pack files at the start. We could indeed move the index files completely to the local side, but that would break backwards compatibility with older versions of restic. Besides, that would still require listing all pack files at each start to find out about removed and new pack files. Depending on the backend, that's a much more expensive operation (in terms of latency and time) than downloading index files.

I think the index file handling in restic has huge potential for optimizations. Let's try implementing an upper bound for the index file size first (since that's backwards compatible) and see where this gets us.

**Comment by rawtaz** [comment]
A question on the thought of not storing the index on the server anymore; Won't this cause practical issues, at least huge performance problems, for the users that for whatever reason don't want to store the cache locally (i.e. that run without the cache)?

As the cache can grow insanely right now, and even if it doesn't, will take up GBs on disk, it's not always wanted to keep a cache local. Would be a bad thing if you'd be left without a cache at all. Then again perhaps that's not what we're talking about here.

**Comment by middelink** [comment]
rawtaz@ i think that question is answered by fd0 in the above 2nd paragraph :) and yes.

On #979 I posted the idea of using a bloomfilter, so that backup only, small machines can use only the bloom filter to determine if the hash is there. We would need some way of appending/joining to the index though, for the restic commands which need access to the actual data. But a bloomfilter can be a few orders of magnitude smaller than the complete index.

**Comment by fd0** [comment]
I've implemented this idea in #1584, can you test it and report back?

**Comment by zcalusic** [comment]
Sure thing!

Both versions compiled with go1.10beta2

**1st run, v0.8.1-92-g164acceb, baseline**

single index, 371MB
3091100kB restic prune

**2nd run, v0.8.1-94-g739bf8e3, #1584**

in the end, index split to 8 indexes
not counting the smallest one, average size is 49MB
2297460kB restic prune

**3rd run, v0.8.1-94-g739bf8e3, #1584**

2930956kB restic prune

I'd say it works well, even if there's no apparent improvement in memory consumption, as you expected.

At this point, seeing how easy it was to split the index, I even wonder how about setting average index file size to match the size of data blobs (4.59M in this repo)? That would lead to about 80 index files in this metadata heavy repo, I feel it would work very well.

**Comment by fd0** [comment]
Thanks for the numbers! For just looking at the impact, `rebuild-index` or even `backup` is probably better.

I'd like to not make the index files too small, then the overhead of downloading (or reading them from local cache) becomes bigger.

The next step here is to reduce the concurrency when reading the index files at the start of the backup (20 or so at the moment), then there won't be as much allocations in parallel at start, which helps memory usage for backup.
